src/passes/stringify-walker-impl.h

```diff
@@ -0,0 +1,120 @@
+#include "stringify-walker.h"
+
+#ifndef wasm_passes_stringify_walker_impl_h
+#define wasm_passes_stringify_walker_impl_h
+
+namespace wasm {
+
+template<typename SubType>
+inline void StringifyWalker<SubType>::doWalkModule(Module* module) {
+  ModuleUtils::iterDefinedFunctions(*module, [&](Function* func) {
+    this->walkFunction(func);
+  });
+}
+
+template<typename SubType>
+inline void StringifyWalker<SubType>::doWalkFunction(Function* func) {
+  walk(func->body);
+  addUniqueSymbol();
+}
+
+template<typename SubType>
+inline void StringifyWalker<SubType>::walk(Expression* curr) {
+  Super::walk(curr);
+  do {
+    addUniqueSymbol();
+    dequeueControlFlow();
+  } while (!controlFlowQueue.empty());
+}
+
+template<typename SubType>
+inline void StringifyWalker<SubType>::scan(SubType* self, Expression** currp) {
+  Expression* curr = *currp;
+  if (Properties::isControlFlowStructure(curr)) {
+    self->controlFlowQueue.push(currp);
+    self->pushTask(doVisitExpression, currp);
+    if (auto* iff = curr->dynCast<If>()) {
+      Super::scan(self, &iff->condition);
+    }
+  } else {
+    Super::scan(self, currp);
+    return;
+  }
+}
+
+/*
+ * This dequeueControlFlow is responsible for ensuring the children expressions
+ * of control flow expressions are visited after the control flow expression has
+ * already been visited. In order to perform this responsibility, the
+ * dequeueControlFlow function needs to always be the very last task in the
+ * Walker stack, as the last task will be executed last. This way if the queue
+ * is not empty, the first statement pushes a new task to call
+ * dequeueControlFlow again.
+ *
+ */
+template<typename SubType> void StringifyWalker<SubType>::dequeueControlFlow() {
+  auto& queue = this->controlFlowQueue;
+  if (queue.empty()) {
+    return;
+  }
+
+  Expression** currp = queue.front();
+  queue.pop();
+  auto self = static_cast<SubType*>(this);
+  Expression* curr = *currp;
+
+  switch (curr->_id) {
+    case Expression::Id::BlockId: {
+      auto* block = curr->cast<Block>();
+      for (auto& child : block->list) {
+        Super::walk(child);
+      }
+      break;
+    }
+    case Expression::Id::IfId: {
+      auto* iff = curr->cast<If>();
+      Super::walk(iff->ifTrue);
+      if (iff->ifFalse) {
+        self->addUniqueSymbol();
+        Super::walk(iff->ifFalse);
+      }
+      break;
+    }
+    case Expression::Id::TryId: {
+      auto* tryy = curr->cast<Try>();
+      Super::walk(tryy->body);
+      for (auto& child : tryy->catchBodies) {
+        this->addUniqueSymbol();
+        Super::walk(child);
+      }
+      break;
+    }
+    case Expression::Id::LoopId: {
+      auto* loop = curr->cast<Loop>();
+      Super::walk(loop->body);
+      break;
+    }
```

Comment:
Maybe we can add a ControlChildIterator to parallel ValueChildIterator, that iterates on control flow children, and use that here? It would avoid us needing to update this code each time we add or modify a control flow structure (i.e. it would keep that logic in `iterator.h`.

I mean, in the code here, we could have:
```cpp
for (auto* child : ControlChildIterator(curr)) {
  this->addUniqueSymbol();
  Super::walk(child);
}
```
However that won't work for block, I guess... so we need some special casing here anyhow. Maybe it's not worth it then.

Another option might be `BlockChildIterator` or such, as what we look for here is a child that is a new block scope..?

Reply:
I think this is worth adding but would incur quite a bit of boilerplate. I've added a TODO and would like to get to it in a follow-on PR.

Reply:
Sounds good. It's not urgent, just an idea (which might not work well, I'm not sure).